From 2ca24dfd4cdf0fa4ed723b65cb4841725178ce42 Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Fri, 17 Oct 2025 17:35:18 -0500
Subject: [PATCH 26/48] ntdll: Return FILE_ATTRIBUTE_REPARSE_POINT from
 get_file_info().

(cherry picked from commit 4e27dcb0b53a9805720f4ef6dd888c60c24af5ef)

CW-Bug-Id: #26492
---
 dlls/ntdll/tests/file.c | 4 ++--
 dlls/ntdll/unix/file.c  | 6 +++++-
 2 files changed, 7 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/tests/file.c b/dlls/ntdll/tests/file.c
index ed368a10b23..6c20c99dd43 100644
--- a/dlls/ntdll/tests/file.c
+++ b/dlls/ntdll/tests/file.c
@@ -6436,7 +6436,7 @@ static void test_reparse_points(void)
     ret = GetFileAttributesW( L"file" );
     ok( ret == FILE_ATTRIBUTE_ARCHIVE, "got %#x\n", ret );
     ret = GetFileAttributesW( L"." );
-    todo_wine ok( ret == (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_REPARSE_POINT), "got %#x\n", ret );
+    ok( ret == (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_REPARSE_POINT), "got %#x\n", ret );
 
     SetCurrentDirectoryW( path2 );
 
@@ -6449,7 +6449,7 @@ static void test_reparse_points(void)
     swprintf( path, ARRAY_SIZE(path), L"%s/testreparse_dirlink", temp_path );
     find_handle = FindFirstFileW( path, &find_data );
     ok( find_handle != INVALID_HANDLE_VALUE, "got error %lu\n", GetLastError() );
-    todo_wine ok( find_data.dwFileAttributes == (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_REPARSE_POINT),
+    ok( find_data.dwFileAttributes == (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_REPARSE_POINT),
         "got attributes %#lx\n", find_data.dwFileAttributes );
     todo_wine ok( find_data.dwReserved0 == IO_REPARSE_TAG_MOUNT_POINT, "got tag %#lx\n", find_data.dwReserved0 );
     FindClose( find_handle );
diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 2a2620e4c21..49d9f268175 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -1785,6 +1785,7 @@ static NTSTATUS fd_set_file_info( int fd, HANDLE handle, UINT attr, BOOL force_s
 /* get the stat info and file attributes for a file (by name) */
 static int get_file_info( const char *path, struct stat *st, ULONG *attr )
 {
+    size_t len = strlen( path );
     char *parent_path;
     char attr_data[65];
     int attr_len, ret;
@@ -1799,7 +1800,7 @@ static int get_file_info( const char *path, struct stat *st, ULONG *attr )
         /* is a symbolic link and a directory, consider these "reparse points" */
         if (S_ISDIR( st->st_mode )) *attr |= FILE_ATTRIBUTE_REPARSE_POINT;
     }
-    else if (S_ISDIR( st->st_mode ) && (parent_path = malloc( strlen(path) + 4 )))
+    else if (S_ISDIR( st->st_mode ) && (parent_path = malloc( len + 4 )))
     {
         struct stat parent_st;
 
@@ -1814,6 +1815,9 @@ static int get_file_info( const char *path, struct stat *st, ULONG *attr )
     }
     *attr |= get_file_attributes( st );
 
+    if (path[len - 1] == '?')
+        *attr |= FILE_ATTRIBUTE_REPARSE_POINT;
+
     attr_len = xattr_get( path, SAMBA_XATTR_DOS_ATTRIB, attr_data, sizeof(attr_data)-1 );
     if (attr_len != -1)
         *attr |= parse_samba_dos_attrib_data( attr_data, attr_len );
-- 
2.53.0

